[종근] 배민찬 1단계 - 회원가입, 로그인 구현

Bellroute opened the pull request.

Bellroute committed:
feat: create Account Acceptance test

Bellroute committed:
feat: sign-up

Bellroute committed:
feat: Account Validation

Bellroute committed:
feat: login function

Bellroute committed:
feat: password-encoder

Bellroute committed:
refactor: password-encoder

Bellroute committed:
chore: 예외처리

Bellroute committed:
feat: front ajax (ing)

Bellroute committed:
refactor: frontend

Bellroute committed:
feat: 회원가입, 로그인 예외 처리

Bellroute committed:
refactor: 예외 처리

Bellroute committed:
chore: add gitignore

Bellroute committed:
chore: add gitignore (.DS_Store)

hyukjin-lee commented on src/main/java/codesquad/domain/Account.java:
```diff
@@ -0,0 +1,65 @@
+@NoArgsConstructor
+public class Account {
+    public static final int USER = 0;
+    public static final int ADMIN = 1;
```
Enum type 으로 빼는게 좋지 않을까요

Bellroute replied:
Enum이 아직 생소해서 생각을 못했네요. 공부해서 반영해보겠습니다

hyukjin-lee commented on src/main/java/codesquad/domain/Account.java:
```diff
@@ -0,0 +1,65 @@
+    private String name;
+
+    @NotBlank
+    @Pattern(regexp = "^\\d{3}-\\d{3,4}-\\d{4}$", message = "비밀번호 형식이 올바르지 않습니다.")
```
messageSourceAccesor를 적용해서 메시지를 설정파일로 관리해보세요

Bellroute replied:
학습 후 반영하겠습니다

hyukjin-lee commented on src/main/java/codesquad/domain/Account.java:
```diff
@@ -0,0 +1,65 @@
+
+@Entity
+@Getter
+@Setter
```
setter는 어느경우에 필요한가요?

Bellroute replied:
혹시 테스트코드에서 사용 되지 않을까 싶어서 추가 했었을텐데 그냥 습관처럼 추가 했던거 같습니다.

hyukjin-lee commented on src/main/java/codesquad/dto/SignUpDTO.java:
```diff
@@ -0,0 +1,43 @@
+    private String email;
+
+    @NotBlank(message = "비밀번호를 입력해 주세요.")
+    @Pattern(regexp = "^(?=.*[A-Za-z])(?=.*\\d)[A-Za-z\\d]{8,16}$", message = "비밀번호 형식이 올바르지 않습니다.")
```
비밀번호를 사용하는 객체가 많아지면 개발하는 사람마다 서로 다른 정책을 적용할 수 있으니
중복을 줄일 필요성이 있어보입니다. 어떻게 하면 될까요?

Bellroute replied:
유틸클래스를 하나 만들어 regex를 상수화시켜 관리하는 방법은 어떤가요?

hyukjin-lee commented on src/main/java/codesquad/dto/SignUpDTO.java:
```diff
@@ -0,0 +1,43 @@
+    private String phoneNumber;
+
+    @Size(min = 2, max = 15, message = "이름은 2~15자 이내로 작성해 주세요.")
+    @Column(nullable = false, name = "name")
```
DTO 는 엔티티가 아닌데 Column 을 붙여주신 이유가 뭔가요?

Bellroute replied:
모델의 validation을 DTO에도 반영하는 과정에서 실수한 것 같습니다. 반영하겠습니다

hyukjin-lee commented on src/main/java/codesquad/dto/SignUpDTO.java:
```diff
@@ -0,0 +1,43 @@
+    private String passwordForCheck;
+
+    @NotBlank(message = "전화번호를 입력해 주세요.")
+    @Pattern(regexp = "^\\d{3}-\\d{3,4}-\\d{4}$", message = "전화번호 형식이 올바르지 않습니다.")
```
DTO에도 철저하게 validation 을 넣어주셨네요.
DTO에도 검증 절차를 넣으면 어떤 이점이 있고 어떤 단점이 있을까요

Bellroute replied:
DTO에 검증 절차 안넣으면 그냥 에러가 안걸리던데 문제있는 건가요?

hyukjin-lee replied:
여기에 적어준 이유를 알고 쓴건지 궁금해서 여쭤봤습니다.
혹시 이게 어떻게 동작하는지 아시나요?
hibernate의 bean validation은 
@Validated, @Valid, validator.validate() 를 호출 할 때 유효성 검사를 하는건데

@Entity 를 붙인 클래스를 유효성 검사하는건 이 코드에 없는 것 같은데
왜 붙이는 것 일까요? 

의문이 생겨서 찾아보니 JPA, Hibernate ORM을 사용하는 경우에는 엔티티를 영속화하거나 업데이트 할 때 validate를 내부적으로 한번 돌린다고 합니다. 
![image](https://user-images.githubusercontent.com/14870706/61924116-632d0100-afa1-11e9-8761-ce9b6fd368c3.png)

그러면 결론적으로 이 코드는 DTO에서 한번 @Valid를 통해 validate를 하고, Entity에서도 저장할 때 validate을 한번 더 하는 것인데, 왜 이렇게 두번이나 하게 구현해야하는 것 일까요?

hyukjin-lee commented on src/main/java/codesquad/exception/UnAuthenticationException.java:
```diff
@@ -0,0 +1,16 @@
+    public UnAuthenticationException(String message) {
+        super(message);
+    }
+}
```
controller advice 를 적용해보세요 라고 했었습니다

hyukjin-lee replied:
aop 에 대한 기본 개념을 잡고 controller advice 를 적용해보세요

Bellroute replied:
aop 공부하고 controller advice 반영하겠습니다

hyukjin-lee commented on src/main/java/codesquad/service/AccountService.java:
```diff
@@ -0,0 +1,68 @@
+
+@Service
+public class AccountService {
+    private static final String SESSIONED_USER = "sessionedUser";
```
세션이라 글로벌하게 쓰일 것 같은데 서비스에 두는게 최선일까요?

hyukjin-lee commented on src/main/java/codesquad/service/AccountService.java:
```diff
@@ -0,0 +1,68 @@
+    private AccountRepository accountRepository;
+
+    @Autowired
+    private PasswordEncoder passwordEncoder;
```
앞으로 스터디에서 Autowired 를 사용하지 말아봅시다

Bellroute replied:
Autowired를 왜 쓰면 안되는지 궁금합니다!

hyukjin-lee replied:
https://www.youtube.com/watch?v=qmI7uJapocw

저번에 창환이가 꼭 보라고 올려준 것인데 참고 ~

hyukjin-lee commented on src/main/java/codesquad/service/AccountService.java:
```diff
@@ -0,0 +1,68 @@
+
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.APPLICATION_JSON);
+        return new ResponseEntity<>(headers, HttpStatus.CREATED);
```
service 레이어는 Http 에 의존성을 갖지 않는게 낫지 않을까요?
(service 레이어가 Http를 모르게 설계하는게 낫지 않을까요?)
이것에 대해 곰곰히 생각해보고 답글 주세요

hyukjin-lee commented on src/main/java/codesquad/service/AccountService.java:
```diff
@@ -0,0 +1,68 @@
+    public ResponseEntity<Account> login(HttpSession session, LoginDTO loginDTO) {
+        Account account = accountRepository.findByEmail(loginDTO.getEmail()).orElseThrow(NotFoundAccountException::new);
+
+        if (!passwordEncoder.matches(loginDTO.getPassword(), account.getPassword())) {
```
둘다 인코딩 된 패스워드인데 passwordEncoder 로 매치 시키는 이유가 있나요? 궁금해서

Bellroute replied:
loginDTO에는 사용자가 로그인시 입력한 값이어서 인코딩 전 패스워드라 매치했습니다.

hyukjin-lee commented on src/main/java/codesquad/web/ApiAccountController.java:
```diff
@@ -0,0 +1,36 @@
+    @Autowired
+    private AccountService accountService;
+
+    @PostMapping("/member/sign-up")
```
- 는 안쓰는게 좋습니다

Bellroute replied:
넵

hyukjin-lee commented on src/main/resources/static/js/script.js:
```diff
@@ -0,0 +1,122 @@
+
+function redirectLoginForm(data) {
+    console.log(data);
+    window.location = 'http://localhost:8080/login';
```
이렇게 하는게 맞을까요? 뭐가 문젠지 말하고 한번 고쳐보세요

Bellroute replied:
정적(?)으로 구현해서 ul이 바뀌면 문제가 생길 것 같습니다. response에 리다이렉트 url을 함께 담아서 보내줘야 할까요?

hyukjin-lee replied:
localhost:8080은 현재 로컬에서만 이 서비스를 사용하겠다는 뜻인데..
https://4rgos.tistory.com/17
참고해서 상대경로로 잘 구현해보세요

hyukjin-lee commented on src/test/java/codesquad/config/PasswordEncoderTest.java:
```diff
@@ -0,0 +1,27 @@
+
+        assertThat(encodedPassword.equals(password)).isFalse();
+        assertThat(passwordEncoder.matches(password, encodedPassword)).isTrue();
+    }
```
이 테스트가 왜 필요 없는지 설명해보세요

Bellroute replied:
이미 잘 구현된 라이브러리라서?
사실 passwordEncoder가 어떻게 돌아가는지 보려고 테스트 짠거였습니다.

hyukjin-lee commented on src/test/java/codesquad/domain/AccountTest.java:
```diff
@@ -0,0 +1,17 @@
+    @Test
+    public void passwordMatch() {
+        String password = "abcd1234";
+        Account account = new Account("be@gmail.com", password, "bell", "010-0000-0000");
```
빌더패턴을 적용해서 가독성있게 구현해보세요

Bellroute replied:
빌더패턴 공부하고 반영하겠습니다.

hyukjin-lee commented on src/test/java/codesquad/domain/AccountTest.java:
```diff
@@ -0,0 +1,17 @@
+        Account account = new Account("be@gmail.com", password, "bell", "010-0000-0000");
+
+        assertThat(account.matchPassword(password)).isTrue();
+        assertThat(account.matchPassword("11111111")).isFalse();
```
이런 경우 assertAll 을 한번 사용해봅시다

Bellroute replied:
assertAll 알아보고 반영하겠습니다.

hyukjin-lee replied:
ㅋㅋㅋㅋ 군대냐;;

hyukjin-lee commented on src/test/java/codesquad/service/AccountServiceTest.java:
```diff
@@ -0,0 +1,99 @@
+
+
+@RunWith(MockitoJUnitRunner.class)
+@SpringBootTest
```
Springboottest 를 붙인 이유가 뭘까요?

Bellroute replied:
passwordEncoder 반영시 이것저것 해보다가 안지웠던 것 같습니다

hyukjin-lee reviewed (changes requested):
구현하느라 고생하셨습니다
모든 댓글에 본인의 생각을 답글로 남겨주세요

Bellroute committed:
refactor: 리뷰반영
1. Enum 사용
2. RegexUtils로 정규식 상수화
3. ControllerAdvice 적용
4. Autowired -> 생성자 주입
5. 프론트 redirect 동적 path 전달
6. Account에 빌더패턴 적용
7. assertAll 적용

Bellroute committed:
refactor: messageSourceAccessor 추가

hyukjin-lee reviewed (approved):
리뷰들 달고 확인버튼 누르는걸 깜빡해서 pending 걸려있었네요;

Bellroute committed:
refactor: validation default message test

Bellroute committed:
fix: messages.properties error
